clash-cores/test/Test/Cores/LineCoding8b10b.hs

```diff
@@ -0,0 +1,115 @@
+-- |
+--   Copyright   :  (C) 2024, QBayLogic B.V.
+--   License     :  BSD2 (see the file LICENSE)
+--   Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+--
+--   8b/10b encoding and decoding tests
+module Test.Cores.LineCoding8b10b where
+
+import Clash.Cores.LineCoding8b10b
+import Clash.Hedgehog.Sized.BitVector
+import qualified Clash.Prelude as C
+import Data.Maybe (isNothing)
+import qualified Hedgehog as H
+import qualified Hedgehog.Gen as Gen
+import Test.Tasty
+import Test.Tasty.Hedgehog
+import Test.Tasty.TH
+import Prelude
+
+-- | Check if a 'BitVector' does not contain a sequence of bits with the same
+--   value for 5 or more indices consecutively
+checkBitSequence :: C.BitVector 10 -> Bool
+checkBitSequence cg =
+  isNothing $
+    C.elemIndex True $
+      C.map (f . C.pack) $
+        C.windows1d C.d5 $
+          C.bv2v cg
+ where
+  f a = a == 0b11111 || a == 0b00000
+
+-- | Function that creates a list with a given range that contains a list of
+--   running disparities and 'Symbol8b10b's
+genSymbol8b10bs :: H.Range Int -> H.Gen [(Bool, Symbol8b10b)]
+genSymbol8b10bs range = do
+  n <- Gen.int range
+  genSymbol8b10bs1 n False
+
+-- | Recursive function to generate a list of 'Symbol8b10b's with the correct
+--   running disparity
+genSymbol8b10bs1 :: Int -> Bool -> H.Gen [(Bool, Symbol8b10b)]
+genSymbol8b10bs1 0 _ = pure []
+genSymbol8b10bs1 n rd = do
+  (rdNew, dw) <- genSymbol8b10b rd
+  ((rdNew, dw) :) <$> genSymbol8b10bs1 (pred n) rdNew
+
+-- | Generate a 'Symbol8b10b' by creating a 'BitVector' of length 10 and
+--   decoding it with the 'decode8b10b' function
+genSymbol8b10b :: Bool -> H.Gen (Bool, Symbol8b10b)
+genSymbol8b10b rd = Gen.filter f $ decode8b10b rd <$> genDefinedBitVector
+ where
+  f (_, dw) = isDw dw
+
+-- Check if the output of 'decode8b10b' is a valid value for a given value from
+-- 'encode8b10b'.
+prop_decode8b10bCheckNothing :: H.Property
+prop_decode8b10bCheckNothing = H.withTests 1000 $ H.property $ do
+  inp <- H.forAll genDefinedBitVector
+  let out = isValidSymbol dw1 && isValidSymbol dw2
+       where
+        (_, dw1) = decode8b10b False $ snd $ encode8b10b False (Dw inp)
+        (_, dw2) = decode8b10b True $ snd $ encode8b10b True (Dw inp)
+
+  H.assert out
```

Comment:
```suggestion
  H.assert $ isValidSymbol $ snd $ decode8b10b False $ snd $ encode8b10b False (Dw inp)
  H.assert $ isValidSymbol $ snd $ decode8b10b True $ snd $ encode8b10b True (Dw inp)
```

The nice thing about this is the formatting of a counter-example. Suppose I corrupt the test by deliberately mismatching the running disparity, Tasty will tell me:

```
peter@gronau:~/clash/clash-cores$ cabal run unittests -- -p LineC -j12
Unittests
  Test.Cores.LineCoding8b10b
    decode8b10bCheckNothing: FAIL (0.03s)
        ✗ <interactive> failed at test/Test/Cores/LineCoding8b10b.hs:59:3
          after 1 test and 2 shrinks.
          shrink path: 1:a2
        
             ┏━━ test/Test/Cores/LineCoding8b10b.hs ━━━
          56 ┃ prop_decode8b10bCheckNothing :: H.Property
          57 ┃ prop_decode8b10bCheckNothing = H.withTests 1000 $ H.property $ do
          58 ┃   inp <- H.forAll genDefinedBitVector
             ┃   │ 0 b0000_0000
          59 ┃   H.assert $ isValidSymbol $ snd $ decode8b10b True $ snd $ encode8b10b False (Dw inp)
             ┃   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
          60 ┃   H.assert $ isValidSymbol $ snd $ decode8b10b True $ snd $ encode8b10b True (Dw inp)
        
          This failure can be reproduced by running:
          > recheckAt (Seed 17468430201590260638 5530981405269087381) "1:a2" <property>
        
      Use "--pattern '$NF ~ /decode8b10bCheckNothing/' --hedgehog-replay '1:a2 Seed 17468430201590260638 5530981405269087381'" to reproduce from the command-line.
      
      Use -p '/LineC/&&/decode8b10bCheckNothing/' to rerun this test only.
```

It pinpoints the exact test that failed.

Reply:
Nice, I wasn't aware this was possible. Thanks!